node/src/components/storage.rs

```diff
@@ -1616,86 +1579,194 @@ fn initialize_block_body_v1_db(
+    block_body_hash: &Digest,
+) -> Result<Option<BlockBody>, LmdbExtError> {
+    let deploy_hashes_with_proof: MerkleLinkedListNode<Vec<DeployHash>> =
+        match get_merkle_linked_list_node(tx, block_body_v2_db, deploy_hashes_db, block_body_hash)?
+        {
+            Some(deploy_hashes_with_proof) => deploy_hashes_with_proof,
+            None => return Ok(None),
+        };
+    let transfer_hashes_with_proof: MerkleLinkedListNode<Vec<DeployHash>> =
+        match get_merkle_linked_list_node(
+            tx,
+            block_body_v2_db,
+            transfer_hashes_db,
+            deploy_hashes_with_proof.merkle_proof_of_rest(),
+        )? {
+            Some(transfer_hashes_with_proof) => transfer_hashes_with_proof,
+            None => return Ok(None),
+        };
+    let proposer_with_proof: MerkleLinkedListNode<PublicKey> = match get_merkle_linked_list_node(
         tx,
         block_body_v2_db,
-        transfer_hashes_db,
-        digest_key_to_delete,
-    )?;
-    // Don't delete proposer since it might be shared
-    Ok(())
+        proposer_db,
+        transfer_hashes_with_proof.merkle_proof_of_rest(),
+    )? {
+        Some(proposer_with_proof) => {
+            debug_assert_eq!(*proposer_with_proof.merkle_proof_of_rest(), hash::SENTINEL1);
+            proposer_with_proof
+        }
+        None => return Ok(None),
+    };
+    let block_body = BlockBody::new(
+        proposer_with_proof.take_value(),
+        deploy_hashes_with_proof.take_value(),
+        transfer_hashes_with_proof.take_value(),
+    );
+
+    Ok(Some(block_body))
 }
 
 /// Purges stale entries from the (Merklized) block body database, and checks the integrity of the
 /// remainder if `should_check_integrity` is true.
 fn initialize_block_body_v2_db(
     env: &Environment,
+    block_header_db: &Database,
     block_body_v2_db: &Database,
     deploy_hashes_db: &Database,
     transfer_hashes_db: &Database,
-    _proposer_db: &Database,
-    deleted_block_body_hashes: HashSet<Digest>,
-    // should_check_integrity: bool,
+    proposer_db: &Database,
+    should_check_integrity: bool,
 ) -> Result<(), LmdbExtError> {
     info!("initializing v2 block body database");
+
     let mut txn = env.begin_rw_txn()?;
-    for digest_key_to_delete in deleted_block_body_hashes {
-        info!(?digest_key_to_delete, "deleting v2 block body");
-        delete_single_block_body_v2(
-            &mut txn,
-            block_body_v2_db,
-            deploy_hashes_db,
-            transfer_hashes_db,
-            digest_key_to_delete,
-        )?;
+
+    let block_body_hash_to_header_map =
+        construct_block_body_to_block_header_reverse_lookup(&txn, block_header_db)?;
+
+    // This will store all the keys that are reachable from a block header, in all the parts
+    // databases (we're basically doing a mark-and-sweep below).
+    let mut live_digests = HashSet::new();
+
+    for (body_hash, header) in &block_body_hash_to_header_map {
```

Comment:
Do we not have the same problem here as we do with V1 version? Where the same part can be referenced from other parts?

Reply:
We do, but that's why this loop only collects the "live" digests (so the entries that are referenced by some header, directly or indirectly), and only the "non-live" ones are deleted later.